gnomad/utils/constraint.py

```diff
@@ -700,3 +701,364 @@ def get_all_pop_lengths(
+    obs: hl.expr.NumericExpression,
+    exp: hl.expr.NumericExpression,
+    output: str = "z_raw",
+) -> hl.Table:
+    """
+    Compute the signed raw z score using observed and expected variant counts.
+
+    The raw z scores are positive when the transcript had fewer variants than expected, and are negative when transcripts had more variants than expected.
+
+    The following annotation is included in the output Table in addition to the `input_ht` keys:
+        - `output` - the raw z score
+
+    :param input_ht: Input Table.
+    :param obs: Observed variant count expression.
+    :param exp: Expected variant count expression.
+    :param output: The annotation label to use for the raw z score output, defaults to 'z_raw'.
+    :return: Table with raw z scores.
+    """
+    ht = input_ht.select(_obs=obs, _exp=exp)
+    ht = ht.annotate(_chisq=(ht._obs - ht._exp) ** 2 / ht._exp)
+    return ht.select(
+        **{output: hl.sqrt(ht._chisq) * hl.if_else(ht._obs > ht._exp, -1, 1)}
+    )
+
+
+def calculate_z_scores(ht: hl.Table, mutation_type) -> hl.Table:
+    """
+    Calculate z scores for synomynous variants, missense variants, and pLoF variants.
+
+    z score = {variant_annotation}_z_raw / {variant_annotation}_sd (variant_annotation could be syn, mis, or lof)
+
+    Function will add the following annotations to output Table:
+        - {mutation_type}_sd (global) - standard deviation of raw z score
+        - constraint_flag - Reason gene does not have constraint metrics. One of:
+            no variants: Zero observed synonymous, missense, pLoF variants
+            no_exp_syn: Zero expected synonymous variants
+            no_exp_mis: Zero expected missense variants
+            no_exp_lof: Zero expected pLoF variants
+            syn_outlier: Too many or too few synonymous variants; synonymous z score < -5 or synonymous z score > 5
+            mis_too_many: Too many missense variants; missense z score < -5
+            lof_too_many: Too many pLoF variants; pLoF z score < -5
+        - syn_z - z score of synonymous variants
+        - mis_z - z score of missense variants
+        - lof_z - z score of pLoF variants
+
+    :param ht: Input Table with observed and expected variant counts for one of
+        synomynous variants, missense variants, and pLoF variants.
+    :param mutation_type: The mutation type of variants in the input Table, one of
+        'lof', 'mis', and 'syn'.
+    :return: Table with z scores.
+    """
+    # Calculate z scores.
+    ht = ht.annotate(
+        **calculate_z(
+            ht,
+            ht[f"obs_{mutation_type}"],
+            ht[f"exp_{mutation_type}"],
+            f"{mutation_type}_z_raw",
+        )[ht.key]
+    )
+
+    # Create constraint flag to annotate if the variant is defined, has expected variant count, and is outlier.
+    reasons = hl.empty_set(hl.tstr)
+
+    reasons = hl.if_else(
+        hl.or_else(ht[f"obs_{mutation_type}"], 0) == 0,
+        reasons.add("no_variants"),
```

Comment:
The old code was:
```
reasons = hl.cond(hl.or_else(ht.obs_syn, 0) + hl.or_else(ht.obs_mis, 0) + hl.or_else(ht.obs_lof, 0) == 0, reasons.add('no_variants'), reasons)
```
My understanding is that "no_variants" will be added to variants with `hl.or_else(ht.obs_[the other two], 0) !=0` in this case, which is fine because they won't be used in the calculation. I'm not 100% confident. Please correct me if I'm wrong.